### PR TITLE
SE-3 | Added functionality to add/update international addresses

### PR DESCRIPTION
**Pull request by periklas**

### CODE REVIEW

File: app/models/dto/AddressDto.scala [hunk 1]
```diff
@@ -58,11 +58,38 @@ object AddressDto {
...
+        .verifying("error.line3_invalid_characters", e => validateAddressLineCharacters(e)),
+      "line4" -> optional(text)
+        .verifying("error.line4_contains_more_than_35_characters", e => e.fold(true)(_.length <= 35))
+        .verifying("error.line4_invalid_characters", e => validateAddressLineCharacters(e)),
+      "line5" -> optional(text)
+        .verifying("error.line5_contains_more_than_35_characters", e => e.fold(true)(_.length <= 35))
+        .verifying("error.line5_invalid_characters", e => validateAddressLineCharacters(e)),
+      "postcode" -> optional(text),
+      "country" -> optional(text)
+        .verifying("error.country_required", _.size > 0),
```
Comment (jennygrahamjones) [inline app/models/dto/AddressDto.scala]: Should the validation here not be a bit stronger? e.g. comparing against the possible countries we will accept.

File: app/views/cards/personaldetails/mainAddress.scala.html [hunk 2]
```diff
@@ -28,7 +28,8 @@
       @address.lines.map { line =>
         @upperCaseToTitleCase(line)<br>
       }
-    @address.postcode.map(_.toUpperCase)
+    @address.postcode.map(_.toUpperCase)<br>
```
Comment (Fenwick17) [inline app/views/cards/personaldetails/mainAddress.scala.html]: Don't use `<br>`

File: app/views/cards/personaldetails/mainAddress.scala.html [hunk 3]
```diff
@@ -28,7 +28,8 @@
       @address.lines.map { line =>
         @upperCaseToTitleCase(line)<br>
```
Comment (Fenwick17) [inline app/views/cards/personaldetails/mainAddress.scala.html]: Don't use `<br>`

File: app/views/cards/personaldetails/mainAddress.scala.html [hunk 4]
```diff
@@ -28,7 +28,8 @@
...
       }
-    @address.postcode.map(_.toUpperCase)
+    @address.postcode.map(_.toUpperCase)<br>
+    @if(!countries.contains(Country(address.country.map(_.toUpperCase).getOrElse("")))) {@address.country.map(_.toUpperCase)}
   }
   </strong></p>
```
Comment (Fenwick17) [inline app/views/cards/personaldetails/mainAddress.scala.html]: `<strong>` has semantic meaning. Is this only applied to provide bold? Or is it an important message

File: app/views/personaldetails/reviewChanges.scala.html [hunk 5]
```diff
@@ -56,8 +63,15 @@ <h1 class="heading-xlarge no-margin-bottom">@Messages("label.check_your_details"
               }
             )
           </dt>
-          <dd class="cya-answer">@Html(addressDto.toList.mkString(", <br>"))</dd>
-          <dd class="cya-change"><a href="@controllers.routes.AddressController.showUpdateAddressForm(typ)">@Messages("label.change_address")</a></dd>
+
+            <dd class="cya-answer">@Html(addressDto.toList.mkString(", <br>"))</dd>
```
Comment (Fenwick17) [inline app/views/personaldetails/reviewChanges.scala.html]: Don't use `<br>`

File: app/models/dto/AddressDto.scala (hunk 1 above)
Comment (connor-irvine) [reply app/models/dto/AddressDto.scala]: Yeah echoing this one. When autocomplete on chrome adds the United Kingdom sometimes it lets you press continue and changed the country to 'ARAB EMIRATES (UNITED)'.  The only thing that's come out of testing that I'm concerned about. @swaistle

File: app/controllers/helpers/CountryHelper.scala [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+ * limitations under the License.
+ */
+
+package controllers.helpers
+
+import models.Country
+
+class CountryHelper {
+
+  val countries = List(
```
[the remark this answers is not in the document]
Comment (periklas) [reply app/controllers/helpers/CountryHelper.scala]: That has now been fixed!

File: test/controllers/AddressControllerSpec.scala [hunk 7]
```diff
@@ -366,7 +367,7 @@ class AddressControllerSpec extends BaseSpec {
       redirectLocation(await(r)) shouldBe Some("/personal-account/your-address/sole/find-address")
     }
 
-    "redirect to Contact HMRC iform page when supplied with value = No (false)" in new LocalSetup {
+    "redirect to enter your cannot use this service page when value = No (false)" in new LocalSetup {
```
[the remark this answers is not in the document]
Comment (derek-gray) [reply test/controllers/AddressControllerSpec.scala]: This has been resolved.

File: app/controllers/AddressController.scala [hunk 8]
```diff
@@ -345,13 +352,13 @@ class AddressController @Inject() (
...
           payeAccount => personDetails => {
-            AddressDto.form.bindFromRequest.fold(
+            AddressDto.ukForm.bindFromRequest.fold(
               formWithErrors => {
                 Future.successful(BadRequest(views.html.personaldetails.updateAddress(formWithErrors, typ, journeyData.addressFinderDto, journeyData.addressLookupServiceDown, showEnterAddressHeader)))
               },
               addressDto => {
                 cacheSubmittedAddressDto(typ, addressDto) flatMap { _ =>
-                  val postCodeHasChanged = !addressDto.postcode.replace(" ", "").equalsIgnoreCase(personDetails.address.flatMap(_.postcode).getOrElse("").replace(" ", ""))
+                  val postCodeHasChanged = !addressDto.postcode.get.replace(" ", "").equalsIgnoreCase(personDetails.address.flatMap(_.postcode).getOrElse("").replace(" ", ""))
```
[the remark this answers is not in the document]
Comment (derek-gray) [reply app/controllers/AddressController.scala]: This has been resolved.

**Comment by jennygrahamjones** [comment]
LGTM! :+1:

**Pull request merged**